cpp/src/arrow/array/builder_nested.h

```diff
@@ -77,6 +77,55 @@ class ARROW_EXPORT ListBuilder : public ArrayBuilder {
   Status AppendNextOffset();
 };
 
+/// \class LargeListBuilder
+/// \brief Builder class for large variable-length list array value types
+///
+/// To use this class, you must append values to the child array builder and use
+/// the Append function to delimit each distinct list value (once the values
+/// have been appended to the child array) or use the bulk API to append
+/// a sequence of offests and null values.
+///
+/// A note on types.  Per arrow/type.h all types in the c++ implementation are
+/// logical so even though this class always builds list array, this can
+/// represent multiple different logical types.  If no logical type is provided
+/// at construction time, the class defaults to List<T> where t is taken from the
+/// value_builder/values that the object is constructed with.
+class ARROW_EXPORT LargeListBuilder : public ArrayBuilder {
```

Comment:
Instead of copy/pasting the `ListBuilder` implementation, how about factoring it into a templated base class that both concrete implementations can inherit from?